src/mozanalysis/stats/bootstrap.py

```diff
@@ -0,0 +1,290 @@
+import numpy as np
+import pandas as pd
+
+import mozanalysis.stats.summarize_samples as masss
+
+
+# Functions that return highly processed stats
+
+
+def compare_branches(
+    sc, df, col_label, ref_branch_label='control', stat_fn=np.mean,
+    num_samples=10000, threshold_quantile=None,
+    individual_summary_quantiles=masss.default_quantiles,
+    comparative_summary_quantiles=masss.default_quantiles
+):
+    """Jointly sample bootstrapped statistics then compare them.
+
+    Args:
+        sc: The Spark context
+        df: a pandas DataFrame of queried experiment data in the
+            standard format (see `mozanalysis.experiment`).
+        col_label (str): Label for the df column contaning the metric
+            to be analyzed.
+        ref_branch_label (str, optional): String in `df['branch']` that
+            identifies the branch with respect to which we want to
+            calculate uplifts - usually the control branch.
+        stat_fn (func, optional): A function that either:
+            - Aggregates each resampled population to a scalar (e.g.
+                the default, `np.mean`), or
+            - Aggregates each resampled population to a dict of
+                scalars.
+            In both cases, this function must accept a one-dimensional
+            ndarray or pandas Series as its input.
+        num_samples (int, optional): The number of bootstrap iterations
+            to perform.
+        threshold_quantile (float, optional): An optional threshold
+            quantile, above which to discard outliers. E.g. `0.9999`.
+        individual_summary_quantiles (list, optional): Quantiles to
+            determine the confidence bands on individual branch
+            statistics. Change these when making Bonferroni corrections.
+        comparative_summary_quantiles (list, optional): Quantiles to
+            determine the confidence bands on comparative branch
+            statistics (i.e. the change relative to the reference
+            branch, probably the control). Change these when making
+            Bonferroni corrections.
+
+    Returns a dictionary:
+    If `stat_fn` returns a scalar (this is the default), then
+    this function returns a dictionary has the following keys and
+    values:
+        'individual': dictionary mapping each branch name to a pandas
+            Series that holds the expected value for the bootstrapped
+            `stat_fn`, and confidence intervals.
+        'comparative': dictionary mapping each branch name to a pandas
+            Series of summary statistics for the possible uplifts of
+            the bootstrapped `stat_fn` relative to the reference branch.
+
+    Otherwise, when `stat_fn` returns a dict, then this function
+    returns a similar dictionary, except the Series are replaced with
+    DataFrames. Each row in each DataFrame corresponds to one output
+    of `stat_fn`, and is the Series that would be returned if `stat_fn`
+    computed only this statistic.
```

Comment:
Yeah, I went back and forth here between these two options. I didn't firmly settle on a Series being the best option, but I leaned that way because the downstream uses are going to be very different depending on whether `stat_fn` returns a scalar or a dict:

* If it returns a scalar then presumably each branch is going to be represented by a scalar, and the data will be presented as a bar chart or similar, with one bar per branch.
* If it returns a dict then presumably each branch is going to be represented by a list of numbers, and the data will be presented as a line or scatter plot, with one series per branch.

I'm also a bit uncomfortable with the fact that you can't tell whether `stat_fn` returns a scalar or dict without executing it. I toyed with the idea of having separate `compare_branches` functions for scalar and multivalued `stat_fn`s, since they really are a separate style of analysis.

So I don't think that there exists a "generic" downstream use, and "tripping" might be a good thing (the closest we get to type safety? haha). Does this change your view? Maybe (as the representative of the downstream user!) you have viable plans in mind for which my concern does not apply?

Reply:
I mostly worry about interactive users copy-pasting things from different examples.

I think mozreport will always use the complex form so I don't have concerns as an implementer.